### DISCUSSION TITLE
Abort execution

### DISCUSSION THREAD

**Discussion by daniel-schroeder-zeiss** [ClearFoundry/ClearScript#252] [Q&A]
Hi,

is there a good way to abort script execution from within the executed script? I'm looking for a way to implement something similar to nodes process.exit(), but I do not want to kill my whole process but only the currently executing script.

**Comment by ClearScriptLib** [comment]
Hi @daniel-schroeder-zeiss,

One possibility is the [`Interrupt`](https://clearscript.clearfoundry.net/Reference/html/M_Microsoft_ClearScript_ScriptEngine_Interrupt.htm) method. You might use it in an extension method like this:

```C#
public static class ScriptEngineExtensions {
    public static bool ExecuteWithExit(this ScriptEngine engine, string code) {
        engine.Script.exit = new Action(() => ScriptEngine.Current.Interrupt());
        try {
            engine.Execute(code);
            return true;
        }
        catch (ScriptInterruptedException) {
            return false;
        }
        finally {
            ((ScriptObject)engine.Script).DeleteProperty("exit");
        }
    }
}
```

And then:

```C#
engine.AddHostType(typeof(Console));
engine.ExecuteWithExit(@"
    Console.WriteLine('foo');
    exit();
    Console.WriteLine('bar'); // NOT REACHED
");
```

However, we can't guarantee that execution will always stop on the exact line of the `exit` call. In some script engines, the native interruption capability could be _asynchronous_.

Good luck!
